ocaml/src/asd_server.ml

```diff
@@ -410,6 +410,109 @@ module Net_fd = struct
          (Lwt_extra2.copy_using reader writer size)
 end
 
+let _range_validate
+      kv io_sched dir_info return
+      ({ RangeQueryArgs.first;finc;last;reverse;max;},
+       verify_checksum, show_all, prio)
+  =
+  let exists_ok (checksum,blob) =
+    match blob with
+    | Value.Direct v  -> Lwt.return None
+    | Value.OnFs(fnr,size) ->
+       let file_name = DirectoryInfo.get_file_path dir_info fnr in
+       Lwt_unix.file_exists file_name >>= fun ok ->
+       Some ("file_exists",ok) |> Lwt.return
+  in
+  let checksum_ok cv =
+    let (checksum, value) = cv in
+    let algo = Checksum.algo_of checksum in
+    Lwt.catch
+      (fun () ->
+        let hash = Hashes.make_hash algo in
+        match value with
+        | Value.Direct blob ->
+           let open Slice in
+           let () = hash # update_substring blob.buf blob.offset blob.length in
+           let checksum2 = hash # final () in
+           let ok = checksum2 = checksum in
+           let open Checksum.Algo in
+           Some (show algo, ok) |> Lwt.return
+        | Value.OnFs (fnr, size) ->
+           DirectoryInfo.with_blob_fd
+             dir_info fnr
+             (fun fd ->
+               Buffer_pool.with_buffer
+                 Buffer_pool.default_buffer_pool
+                 (fun buffer ->
+                   let reader = Lwt_bytes.read fd in
+                   let writer = hash # update_lwt_bytes_detached
+                   in
+                   Lwt_extra2.copy_using reader writer size buffer)
```

Comment:
This will read and hash the entire blob in slices of 4KB (due to the default buffer pool size).
For efficiency it would be better to read bigger slices, e.g. of 1MB.